File: InsuraTech.Domain/Common/Entity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace InsuraTech.Domain.Common
{
    public abstract class Entity
    {
        public Guid Id { get; set; }
        public DateTime CreatedAt { get; protected set; }
        public DateTime UpdatedAt { get; protected set; }
        public bool IsDeleted { get; set; }
        private int _version {  get; set; }
        public int Version => _version;

        protected Entity() {
            Id = Guid.NewGuid();
            CreatedAt = DateTime.UtcNow;
        }

        public void IncrementVersion () => _version++;

        protected void MarkAsUpdated() => UpdatedAt = DateTime.UtcNow;

        public override bool Equals(object? obj)
        {
            if (obj is not Entity other) return false;
            if (ReferenceEquals(this, other)) return true;
            return Id == other.Id;
        }
        public override int GetHashCode () => Id.GetHashCode();

    }
}

File: backend_v2/src/main/java/com/insuratech/domain/common/Entity.java
package com.insuratech.domain.common;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.Id;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

@Getter
public abstract class Entity {

    @Id
    private String id;
    private Instant createdAt;
    @Setter(AccessLevel.PROTECTED)
    private Instant updatedAt;
    @Setter
    private boolean deleted;
    private int version;

    protected Entity() {
        this.id = UUID.randomUUID().toString();
        this.createdAt = Instant.now();
    }

    public void incrementVersion() {
        this.version++;
    }

    protected void markAsUpdated() {
        this.updatedAt = Instant.now();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Entity other)) return false;
        return Objects.equals(id, other.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }
}
